Fixes #6669: Allow images in deck description

Ekagra opened the pull request:
## Pull Request template

## Purpose / Description
Fixes #6669

## Fixes
https://github.com/ankidroid/Anki-Android/issues/6669

## Approach
I've made a HtmlImageGetter class that displays the html image using drawable and ImageGetter, I've used ic_add_toolbar_icon as placeholder.

## How Has This Been Tested?
<p float="left">
  <img src="https://i.imgur.com/klPfcsG.jpg" width="400" />
  <img src="https://i.imgur.com/UPK24fy.jpeg" width="400" /> 
</p>

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [x] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/HtmlImageGetter.java:
```diff
@@ -0,0 +1,90 @@
+        mTv.setText(spanned);
+    }
+
+    @SuppressWarnings( "deprecation" )
```
I'd prefer not. If we can't avoid using a deprecated method, at least I want us to know that we use deprecated method, so that we can correct later. Did you check whether actually there is nothing better than the dprecated one?

Ekagra replied:
I sought to load images using bitmap, drawable and ImageGetter only and it was giving this warning with all of those during the build so I had no other choice.

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/HtmlImageGetter.java:
```diff
@@ -0,0 +1,90 @@
+import android.widget.ImageView;
+import android.widget.TextView;
+
+public class HtmlImageGetter extends Activity implements ImageGetter {
```
Can I please ask for some javadoc. 
I understand that this gets image from HTML, but I'd still love to understand why it's needed and implemented

Ekagra replied:
I can't find a definitive javadoc but here's android reference to it [ImageGetter](https://developer.android.com/reference/android/text/Html.ImageGetter)

Arthur-Milchior commented on AnkiDroid/src/main/java/com/ichi2/anki/HtmlImageGetter.java:
```diff
@@ -0,0 +1,90 @@
+        d.addLevel(0, 0, empty);
+        d.setBounds(0, 0, empty.getIntrinsicWidth(), empty.getIntrinsicHeight());
+
+        new LoadImage().execute(source, d);
```
What's the point of directly executing an async task ? If you are not running it asynchronously, that seems useless.
I doubt async task is the proper tool anyway, I thought async task where made to run things one after the other, and we probably want to start all download at the same time.

Ekagra replied:
You get `android.os.NetworkOnMainThreadException` when it attempts to perform a networking operation on its main thread

Arthur-Milchior commented:
Hi

Thanks a lot for helping improve ankidroid.

Sadly, right now, unit tests fails. In particular, 

> > Task :AnkiDroid:compileDebugUnitTestJavaWithJavac
/home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/StudyOptionsFragmentTest.java:19: error: cannot find symbol
        Spanned result = StudyOptionsFragment.formatDescription("a<span style=\"color:red\">a=1</span>a");

Did you try running those unit test on your machine before submitting?


The goal of commit title is that someone, in a few years, can look at history, and know easily what a commit do, and why a line of code was added. Sadly, #6699 is not descriptive.
Can I please ask you to rephrase, for example:

> Allow images in deck description
> 
> Fixes #6669 by 


In your screenshot, we see html on the bottom left, I don't expect that what we would like to see actually. Not sure what the ideal result is. I believe that no html would be good, and adding the html back when the user put the focus on it and start to edit. Probably worth doing an extra feature request for that.


I'd like you if you can test with let's say 20 images and check whether it loads normally or slowly. I assume that it will load one image after the other here if no actualy background task is used

Arthur-Milchior commented:
The original bug report mentionned using webview. I don't know enough about it to know the difference, but maybe @david-allison-1 can check whether not using it would be a prolbem

Ekagra commented:
> Hi
> 
> Thanks a lot for helping improve ankidroid.
> 
> Sadly, right now, unit tests fails. In particular,
> 
> > > Task :AnkiDroid:compileDebugUnitTestJavaWithJavac
> > > /home/runner/work/Anki-Android/Anki-Android/AnkiDroid/src/test/java/com/ichi2/anki/StudyOptionsFragmentTest.java:19: error: cannot find symbol
> > > Spanned result = StudyOptionsFragment.formatDescription("a<span style="color:red">a=1a");
> 
> Did you try running those unit test on your machine before submitting?
> 
> The goal of commit title is that someone, in a few years, can look at history, and know easily what a commit do, and why a line of code was added. Sadly, #6699 is not descriptive.
> Can I please ask you to rephrase, for example:
> 
> > Allow images in deck description
> > Fixes #6669 by
> 
> In your screenshot, we see html on the bottom left, I don't expect that what we would like to see actually. Not sure what the ideal result is. I believe that no html would be good, and adding the html back when the user put the focus on it and start to edit. Probably worth doing an extra feature request for that.
> 
> I'd like you if you can test with let's say 20 images and check whether it loads normally or slowly. I assume that it will load one image after the other here if no actualy background task is used

I will surely rephrase the PR. Yes it loads them one after the another, is there any other way we have to load images?

david-allison-1 commented:
> The original bug report mentionned using webview. I don't know enough about it to know the difference, but maybe @david-allison-1 can check whether not using it would be a prolbem

It seems like using a WebView would be the correct decision, but maybe restrict it from internet access until we confirm a need for it. This is one that we should probably check with Anki Desktop's source code, and maybe dae about to ensure that he doesn't have concerns about the implementation.

* We need to consider the user privacy aspects of allowing the WebView to access the internet (even just external images), and especially loading javascript - we need to see what Anki Desktop does here and question whether we should emulate it. It would allow a newly downloaded deck to phone home before the deck has been open.
* We need to handle the WebView `onRenderProcessGone` and associated nastiness.

infinyte7 commented:
@Ekagra Can you try this sample for testing if it is working or not?

It is used for showing update available for decks. As user have to manually check for deck's updates. So, I have implemented simple script for doing that. It is working on AnkiDesktop but due to not having webview, it is not working in AnkiDroid


I have added js script to fetch deck version from GitHub and compare with deck version in deck description. If change then it show update available. It is working on AnkiDesktop. So, webview will be helpful here. 

On desktop it show update available for decks
Download [sample.zip](https://github.com/ankidroid/Anki-Android/files/6136627/sample.zip) and import and test it on AnkiDesktop and AnkiDroid

View [versionPopup.html](https://gist.github.com/infinyte7/57a7df4f140d3cf3c9b5dd4788bca6ce#file-versionpopup-html) in deck description

AnkiDesktop
Updating deck with new content with latest version hide the popup.
<img src="https://user-images.githubusercontent.com/12841290/111067833-f7decc80-8500-11eb-9be6-60caa8be608e.PNG" height=400></img>

AnkiDroid
Also updating deck does no effect. 
<img src="https://user-images.githubusercontent.com/12841290/111068169-9ae41600-8502-11eb-9a79-7fe946aeb00f.PNG" height=400></img>

So, what will be approach?
Implementation of android webview will better option here.

Ekagra commented:
> > The original bug report mentionned using webview. I don't know enough about it to know the difference, but maybe @david-allison-1 can check whether not using it would be a prolbem
> 
> It seems like using a WebView may be the correct decision (depends on what Anki Desktop does), but maybe restrict it from internet access until we confirm a need for it. This is one that we should probably check with Anki Desktop's source code, and maybe dae about to ensure that he doesn't have concerns about the implementation.
> 
> * We need to consider the user privacy aspects of allowing the WebView to access the internet (even just external images), and especially loading javascript - we need to see what Anki Desktop does here and question whether we should emulate it. It would allow a newly downloaded deck to phone home before the deck has been open.
> * We need to handle the WebView `onRenderProcessGone` and associated nastiness.

Alright, so we only have to use webview to load images? You can update about what is the correct way to do it and how the desktop version does it.

Ekagra commented:
> @Ekagra Can you try this sample for testing if it is working or not?
> 
> It is used for showing update available for decks. As user have to manually check for deck's updates. So, I have implemented simple script for doing that. It is working on AnkiDesktop but due to not having webview, it is not working in AnkiDroid.
> 
> I have added js script to fetch deck version from GitHub and compare with deck version in deck description. If there are changes then it show update available.
> 
> On desktop it shows update available for decks
> Download [sample.zip](https://github.com/ankidroid/Anki-Android/files/6136627/sample.zip) and import and test it on AnkiDesktop and AnkiDroid
> 
> View [versionPopup.html](https://gist.github.com/infinyte7/57a7df4f140d3cf3c9b5dd4788bca6ce#file-versionpopup-html) in deck description
> 
> AnkiDesktop
> Updating deck with new content with latest version hide the popup.
> <img alt="" height="400" src="https://user-images.githubusercontent.com/12841290/111067833-f7decc80-8500-11eb-9be6-60caa8be608e.PNG">
> 
> AnkiDroid
> Also updating deck does no effect.
> <img alt="" height="400" src="https://user-images.githubusercontent.com/12841290/111068169-9ae41600-8502-11eb-9a79-7fe946aeb00f.PNG">
> 
> So, what will be approach?
> Implementation of android webview will better option here.

Yeah, I'll check it out and see what needs to be done.

mikehardy commented:
So, to summarize all the feedback:

- it appears Anki Desktop uses a WebView for this, must be confirmed, yes?
- Assuming it uses a WebView we should strive to load in a WebView as well in the same way (same restrictions on content - like, no javascript etc)

dae commented:
Future versions of Anki will block JS from the deck descriptions - the fact that it's currently allowed was just an oversight.

david-allison-1 commented:
> Future versions of Anki will block JS from the deck descriptions - the fact that it's currently allowed was just an oversight.

Thanks! Do you plan to continue to allow external content (`http(s)`) and or local images?

dae commented:
At the moment images are filtered out when the new code is enabled, but I suspect deck authors will ask for them again. Local images are likely not an issue; remote images are potentially a tracking bug - but we allow them on cards already, so special-casing the overview screen doesn't seem to buy us much. An option in the preferences to disable remote images everywhere might be nice at one point

mikehardy commented:
I think we've been quite busy with 2.15 and just not had time to work through this one - unstale-ing for now
